File: sober-level.py
import sys
import os

import cv2
import numpy as np

img_fn = sys.argv[1]
bn, en = os.path.splitext(img_fn)
dst_fn = bn + "_sober_edge-level-his" + en

# 读取图像（建议为灰度图）
img = cv2.imread(img_fn, cv2.IMREAD_UNCHANGED) 
print("shape:" + str(img.shape))
print("dtype:" + str(img.dtype))
if len(img.shape) >= 3 and img.shape[2] > 1:
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)  # 转灰度图
else:
    gray = img

blur1 = cv2.GaussianBlur(gray, (35, 35), 0)

# 计算 Sobel 梯度（x 和 y 方向）
sobel_x = cv2.Sobel(blur1, cv2.CV_64F, 1, 0, ksize=3)
sobel_y = cv2.Sobel(blur1, cv2.CV_64F, 0, 1, ksize=3)

# 计算梯度幅值（与 ImageJ "Find Edges" 一致）
sobel_img = cv2.magnitude(sobel_x, sobel_y)

#低于某阈值的部分清空
sobel_img[sobel_img<40] = 0


ret_img = sobel_img 
# 转换为可显示的 8 位图像
edges = cv2.convertScaleAbs(ret_img)

cv2.imwrite(dst_fn, edges)

File: sober-level.cpp
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>

int main(int argc, char* argv[])
{
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <image_filename>" << std::endl;
        return 1;
    }

    std::string img_fn = argv[1];

    // 分离文件名与扩展名
    size_t dot_pos = img_fn.find_last_of('.');
    std::string bn = (dot_pos == std::string::npos) ? img_fn : img_fn.substr(0, dot_pos);
    std::string en = (dot_pos == std::string::npos) ? "" : img_fn.substr(dot_pos);

    std::string dst_fn = bn + "_sober_edge-level-his" + en;

    // 读取图像（建议灰度图）
    cv::Mat img = cv::imread(img_fn, cv::IMREAD_UNCHANGED);
    if (img.empty()) {
        std::cerr << "Failed to load image: " << img_fn << std::endl;
        return 1;
    }

    std::cout << "shape: " << img.rows << "x" << img.cols 
              << (img.channels() > 1 ? "x" + std::to_string(img.channels()) : "")
              << std::endl;
    std::cout << "dtype: " << img.type() << std::endl;

    cv::Mat gray;
    if (img.channels() > 1) {
        cv::cvtColor(img, gray, cv::COLOR_BGR2GRAY);
    } else {
        gray = img.clone();
    }

    cv::Mat blur1;
    cv::GaussianBlur(gray, blur1, cv::Size(35, 35), 0);

    cv::Mat sobel_x, sobel_y;
    cv::Sobel(blur1, sobel_x, CV_64F, 1, 0, 3);
    cv::Sobel(blur1, sobel_y, CV_64F, 0, 1, 3);

    cv::Mat sobel_img;
    cv::magnitude(sobel_x, sobel_y, sobel_img);

    // 低于阈值的部分清零
    for (int r = 0; r < sobel_img.rows; ++r) {
        double* row = sobel_img.ptr<double>(r);
        for (int c = 0; c < sobel_img.cols; ++c) {
            if (row[c] < 40.0) row[c] = 0.0;
        }
    }

    // 转换为 8位图像
    cv::Mat edges;
    cv::convertScaleAbs(sobel_img, edges);

    if (!cv::imwrite(dst_fn, edges)) {
        std::cerr << "Failed to write output: " << dst_fn << std::endl;
        return 1;
    }

    std::cout << "Saved result to " << dst_fn << std::endl;

    return 0;
}
